File: daemon/src/migrate.ts
import {
  cp,
  mkdir,
  readFile,
  readdir,
  rename,
  rm,
  stat,
  writeFile,
} from 'node:fs/promises'
import path from 'node:path'
import type { BroodmotherConfig } from '@broodmother/types/config'
import type { LoadedConfig } from '@broodmother/config'
import { Git } from '@broodmother/git'
import { PROFILE_FILE, listProfiles } from './profiles'
import { REPOS_DIR, listRepos, repoCheckouts } from '@broodmother/repo'
import { PRIMARY } from '@broodmother/branch'
import { TASK_EXTENSION } from '@broodmother/types/task/schema'
import { listProjects, projectCheckouts } from '@broodmother/project'

const STAGING = '.migrating'
const LEGACY_TASK = '.dream'
const LEGACY_PROFILES = 'profiles'
const LEGACY_REGISTRY = 'projects.json'
/** Repos were called projects, and the folder a project keeps them in said so. */
const LEGACY_REPOS_DIR = '.projects'
/** The profile that takes in projects from a home that never had one. */
const FALLBACK = 'default'

/**
 * The layout before the home was a shelf of profiles: profiles were files in `profiles/`,
 * projects were folders beside it, and a repo was a repository anywhere on the disk that a
 * registry in the project pointed at.
 *
 * Everything moves rather than being copied — a git repository is portable, and moving a
 * whole directory keeps it one — and every checkout is repaired afterwards, because a
 * worktree remembers where its repository was in absolute paths. Nothing is deleted except
 * the registry the repos have replaced, and a home already in the new shape is left
 * exactly as it is.
 *
 * Repos were called projects until they were not, and the folder a project keeps them in was
 * named for it — `.projects/` becomes `.repos/`, and every checkout inside is repaired for
 * the same reason as above.
 */
export async function migrate(
  home: string,
  loaded: LoadedConfig,
): Promise<{ config: BroodmotherConfig; moved: string[] }> {
  const staged = await stageProjects(home)
  await adoptProfiles(home)
  if (staged.length && !(await listProfiles(home)).length)
    await writeProfile(home, FALLBACK)

  const profiles = (await listProfiles(home)).map((profile) => profile.name)
  // A project nobody bound goes to the first profile there is — one of them made it, and the
  // machine has forgotten which.
  const owner = (project: string) =>
    pick(loaded.bindings[path.join(home, project)], profiles) ?? profiles[0] ?? FALLBACK

  const moved: string[] = []
  const paths = new Map<string, string>()
  for (const name of staged) {
    const to = await land(path.join(home, STAGING, name), path.join(home, owner(name)))
    paths.set(path.join(home, name), to)
    moved.push(to)
  }
  await rm(path.join(home, STAGING), { recursive: true, force: true })

  for (const profile of profiles)
    for (const project of await listProjects(path.join(home, profile))) {
      await liftRepos(project.path)
      await liftCheckout(project.path)
      await adoptRepos(project.path)
      await repair(projectCheckouts(project.path))
      await adoptTasks(project.path)
    }
  await rmIfEmpty(path.join(home, LEGACY_REPOS_DIR))

  return { config: rewrite(loaded.config, paths, profiles), moved }
}

/** Every folder in the home that is not a profile is a project from the old layout. Staged
 *  out of the way first, so a profile can take the name a project had. */
async function stageProjects(home: string): Promise<string[]> {
  const entries = await readdir(home, { withFileTypes: true }).catch(() => [])
  const legacy: string[] = []
  for (const entry of entries) {
    if (!entry.isDirectory() || entry.name.startsWith('.')) continue
    if (entry.name === LEGACY_PROFILES) continue
    if (await exists(path.join(home, entry.name, PROFILE_FILE))) continue
    legacy.push(entry.name)
  }
  if (!legacy.length) return []

  const staging = path.join(home, STAGING)
  await mkdir(staging, { recursive: true })
  for (const name of legacy) await rename(path.join(home, name), path.join(staging, name))
  return legacy
}

/** `profiles/ada.json` and the key beside it become the folder `ada/` holds. */
async function adoptProfiles(home: string): Promise<void> {
  const dir = path.join(home, LEGACY_PROFILES)
  const entries = await readdir(dir).catch(() => [])
  for (const entry of entries) {
    if (!entry.endsWith('.json') || entry.startsWith('.')) continue
    const name = entry.slice(0, -'.json'.length)
    const target = path.join(home, name)
    await mkdir(target, { recursive: true })
    await rename(path.join(dir, entry), path.join(target, PROFILE_FILE))
    for (const suffix of ['.key', '.key.pub'])
      await rename(
        path.join(dir, `${name}${suffix}`),
        path.join(target, `profile${suffix}`),
      ).catch(() => {})
    await repointKey(path.join(target, PROFILE_FILE), path.join(dir, `${name}.key`))
  }
  await rm(dir, { recursive: true }).catch(() => {})
}

/** The key moved with the profile, and the profile names it by absolute path. */
async function repointKey(file: string, was: string): Promise<void> {
  const raw = await readFile(file, 'utf8')
    .then(JSON.parse)
    .catch(() => null)
  if (!raw || typeof raw !== 'object' || raw.sshKeyPath !== was) return
  const next = { ...raw, sshKeyPath: file.replace(/\.json$/, '.key') }
  await writeFile(file, `${JSON.stringify(next, null, 2)}\n`, { mode: 0o600 })
}

async function writeProfile(home: string, name: string): Promise<void> {
  await mkdir(path.join(home, name), { recursive: true })
  await writeFile(path.join(home, name, PROFILE_FILE), '{}\n', { mode: 0o600 })
}

/** The staged project, into the profile that owns it. A name already taken there is only ever
 *  a migration that stopped halfway, and neither folder is worth losing to the other. */
async function land(from: string, profile: string): Promise<string> {
  await mkdir(profile, { recursive: true })
  const name = path.basename(from)
  let target = path.join(profile, name)
  for (let n = 2; await exists(target); n++) target = path.join(profile, `${name}-${n}`)
  await rename(from, target)
  return target
}

/**
 * The layout before a project held checkouts: the project folder was the checkout itself. It
 * becomes `local/`, so the branches added later are its peers rather than folders buried
 * inside it. A project already holding one is left exactly as it is.
 */
async function liftCheckout(project: string): Promise<void> {
  const local = path.join(project, PRIMARY)
  if (await exists(local)) return
  const entries = await readdir(project).catch(() => [])
  if (!entries.length) return

  // Staged inside the project so every move stays on one device, then renamed into place.
  const staging = path.join(project, STAGING)
  await rm(staging, { recursive: true, force: true })
  await mkdir(staging, { recursive: true })
  for (const entry of entries) {
    // The repos sit beside the checkouts rather than in one, which is what keeps the
    // sync loop from ever seeing them.
    if (entry === STAGING || entry === REPOS_DIR || entry === LEGACY_REPOS_DIR) continue
    await rename(path.join(project, entry), path.join(staging, entry))
  }
  await rename(staging, local)
}

/** `.projects/` becomes `.repos/`, contents and all. A project that already has the new folder
 *  takes in whatever the old one holds that it does not; anything it does hold stays where
 *  it is, because neither copy is worth losing to the other. Every checkout that moved is
 *  repaired, since its worktrees remember the folder by its old name. */
async function liftRepos(project: string): Promise<void> {
  const was = path.join(project, LEGACY_REPOS_DIR)
  const now = path.join(project, REPOS_DIR)
  if (!(await exists(was))) return
  if (!(await exists(now))) {
    await rename(was, now)
  } else {
    for (const entry of await readdir(was).catch(() => [])) {
      if (await exists(path.join(now, entry))) continue
      await rename(path.join(was, entry), path.join(now, entry))
    }
    await rmIfEmpty(was)
  }
  for (const repo of await listRepos(project)) await repair(repoCheckouts(project, repo.name))
}

/** Every repository the registry pointed at, moved into the project as the repo's own
 *  `local`. A repository that is no longer there is an entry with nothing behind it. */
async function adoptRepos(project: string): Promise<void> {
  const file = path.join(project, REPOS_DIR, LEGACY_REGISTRY)
  const raw = await readFile(file, 'utf8')
    .then(JSON.parse)
    .catch(() => null)
  if (!raw || typeof raw !== 'object') return

  for (const [name, repo] of Object.entries(raw as Record<string, unknown>)) {
    if (typeof repo !== 'string' || !repo) continue
    const checkouts = repoCheckouts(project, name)
    if (repo === checkouts.primary || !(await exists(repo))) continue
    await mkdir(checkouts.worktrees, { recursive: true })
    await move(repo, checkouts.primary)
    await repair(checkouts)
  }
  await rm(file, { force: true })
}

/** Tasks were called dreams, and the name was in the extension every one of them wears.
 *  Renamed in place, in every checkout — git sees the rename and the next sync carries it. */
async function adoptTasks(project: string): Promise<void> {
  const entries = await readdir(project, { withFileTypes: true, recursive: true }).catch(
    () => [],
  )
  for (const entry of entries) {
    if (!entry.isFile() || !entry.name.endsWith(LEGACY_TASK)) continue
    const from = path.join(entry.parentPath, entry.name)
    const to = `${from.slice(0, -LEGACY_TASK.length)}${TASK_EXTENSION}`
    if (await exists(to)) continue
    await rename(from, to)
  }
}

/** A rename across devices is not a rename, and a repository on another volume is an
 *  ordinary place to have kept one. */
async function move(from: string, to: string): Promise<void> {
  try {
    await rename(from, to)
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code !== 'EXDEV') throw error
    await cp(from, to, {
      recursive: true,
      preserveTimestamps: true,
      verbatimSymlinks: true,
    })
    await rm(from, { recursive: true, force: true })
  }
}

/** A worktree records where its repository is, and its repository records where it is —
 *  both in absolute paths that the move just invalidated. */
async function repair(checkouts: { primary: string; worktrees: string }): Promise<void> {
  if (!(await exists(path.join(checkouts.primary, '.git')))) return
  const entries = await readdir(checkouts.worktrees, { withFileTypes: true }).catch(
    () => [],
  )
  const trees = entries
    .filter(
      (entry) =>
        entry.isDirectory() && entry.name !== PRIMARY && !entry.name.startsWith('.'),
    )
    .map((entry) => path.join(checkouts.worktrees, entry.name))
  await new Git(checkouts.primary).run(['worktree', 'repair', ...trees])
}

/** Everything this machine filed under a project path, moved onto the path it now has. */
function rewrite(
  config: BroodmotherConfig,
  paths: Map<string, string>,
  profiles: string[],
): BroodmotherConfig {
  const at = (project: string) => paths.get(project) ?? project
  const rekey = <T>(record: Record<string, T>): Record<string, T> =>
    Object.fromEntries(Object.entries(record).map(([key, value]) => [at(key), value]))

  const projectPath = config.projectPath ? at(config.projectPath) : null
  return {
    ...config,
    projectPath,
    profile:
      pick(projectPath ? path.basename(path.dirname(projectPath)) : null, profiles) ??
      config.profile ??
      profiles[0] ??
      null,
    git: rekey(config.git),
    checkouts: rekey(config.checkouts),
    repo: rekey(config.repo),
    repoBranch: Object.fromEntries(
      Object.entries(config.repoBranch).map(([key, value]) => {
        const cut = key.lastIndexOf('#')
        return cut < 0
          ? [key, value]
          : [`${at(key.slice(0, cut))}${key.slice(cut)}`, value]
      }),
    ),
  }
}

const pick = (name: string | null | undefined, names: string[]) =>
  name && names.includes(name) ? name : null

/** The home's old repo folder, once every repository in it has been moved into a project.
 *  Anything left is something broodmother did not put there. */
async function rmIfEmpty(dir: string): Promise<void> {
  const entries = await readdir(dir).catch(() => ['keep'])
  if (!entries.length) await rm(dir, { recursive: true, force: true })
}

const exists = (target: string) =>
  stat(target).then(
    () => true,
    () => false,
  )

File: daemon-go/internal/migrate/migrate.go
// Package migrate moves a home written in an older layout into the one this daemon reads.
//
// The layout before the home was a shelf of profiles: profiles were files in `profiles/`,
// projects were folders beside it, and a repo was a repository anywhere on the disk that a
// registry in the project pointed at.
//
// Everything moves rather than being copied — a git repository is portable, and moving a whole
// directory keeps it one — and every checkout is repaired afterwards, because a worktree
// remembers where its repository was in absolute paths. Nothing is deleted except the registry
// the repos have replaced, and a home already in the new shape is left exactly as it is.
//
// It runs before anything else reads the home, and it runs on every start: a home already
// migrated costs it a handful of directory reads and no writes at all.
package migrate

import (
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/broodmotherai/broodmother/daemon-go/internal/config"
	"github.com/broodmotherai/broodmother/daemon-go/internal/constants"
	"github.com/broodmotherai/broodmother/daemon-go/internal/jsjson"
	"github.com/broodmotherai/broodmother/daemon-go/internal/profile"
	"github.com/broodmotherai/broodmother/daemon-go/internal/project"
)

const (
	// staging is where a project waits while a profile takes the name it had.
	staging = ".migrating"
	// legacyTask: tasks were called dreams, and the name was in the extension every one wore.
	legacyTask = ".dream"
	// legacyProfiles: a profile was a file in this folder rather than a folder of its own.
	legacyProfiles = "profiles"
	// legacyRegistry: a repo was a repository anywhere on the disk that this file pointed at.
	legacyRegistry = "projects.json"
	// legacyReposDir: repos were called projects, and the folder a project keeps them in said so.
	legacyReposDir = ".projects"
	// legacyAttachments: attachments were the one folder in a project that was not dotted, and
	// skills sat beside the tools they run rather than inside them.
	legacyAttachments = "attachments"
	legacySkills      = ".skills"
	// fallback is the profile that takes in projects from a home that never had one.
	fallback = "default"
)

// Run migrates the home and answers with the config as it now reads. The config moves with the
// folders: everything this machine filed under a project path is filed under the path it has.
func Run(home string, loaded config.Loaded) (config.Config, error) {
	staged, err := stageProjects(home)
	if err != nil {
		return config.Config{}, err
	}
	if err := adoptProfiles(home); err != nil {
		return config.Config{}, err
	}
	if len(staged) > 0 {
		held, err := profile.List(home)
		if err != nil {
			return config.Config{}, err
		}
		if len(held) == 0 {
			if err := writeProfile(home, fallback); err != nil {
				return config.Config{}, err
			}
		}
	}

	held, err := profile.List(home)
	if err != nil {
		return config.Config{}, err
	}
	names := make([]string, 0, len(held))
	for _, one := range held {
		names = append(names, one.Name)
	}

	paths := map[string]string{}
	for _, name := range staged {
		to, err := land(filepath.Join(home, staging, name), filepath.Join(home, owner(loaded, home, name, names)))
		if err != nil {
			return config.Config{}, err
		}
		paths[filepath.Join(home, name)] = to
	}
	if err := os.RemoveAll(filepath.Join(home, staging)); err != nil {
		return config.Config{}, err
	}

	for _, name := range names {
		projects, err := project.List(filepath.Join(home, name))
		if err != nil {
			return config.Config{}, err
		}
		for _, one := range projects {
			if err := liftProject(one.Path); err != nil {
				return config.Config{}, err
			}
		}
	}
	rmIfEmpty(filepath.Join(home, legacyReposDir))

	return rewrite(loaded.Config, paths, names), nil
}

// owner is the profile a staged project lands in. A project nobody bound goes to the first
// profile there is — one of them made it, and the machine has forgotten which.
func owner(loaded config.Loaded, home, name string, names []string) string {
	if held := pick(loaded.Bindings[filepath.Join(home, name)], names); held != "" {
		return held
	}
	if len(names) > 0 {
		return names[0]
	}
	return fallback
}

func pick(name string, names []string) string {
	if name == "" {
		return ""
	}
	for _, one := range names {
		if one == name {
			return name
		}
	}
	return ""
}

// stageProjects: every folder in the home that is not a profile is a project from the old
// layout. Staged out of the way first, so a profile can take the name a project had.
func stageProjects(home string) ([]string, error) {
	entries, err := os.ReadDir(home)
	if err != nil {
		return nil, nil
	}
	var legacy []string
	for _, entry := range entries {
		if !entry.IsDir() || strings.HasPrefix(entry.Name(), ".") {
			continue
		}
		if entry.Name() == legacyProfiles {
			continue
		}
		if exists(filepath.Join(home, entry.Name(), constants.ProfileFile)) {
			continue
		}
		legacy = append(legacy, entry.Name())
	}
	if len(legacy) == 0 {
		return nil, nil
	}

	staged := filepath.Join(home, staging)
	if err := os.MkdirAll(staged, 0o755); err != nil {
		return nil, err
	}
	for _, name := range legacy {
		if err := os.Rename(filepath.Join(home, name), filepath.Join(staged, name)); err != nil {
			return nil, err
		}
	}
	return legacy, nil
}

// adoptProfiles: `profiles/ada.json` and the key beside it become the folder `ada/` holds.
func adoptProfiles(home string) error {
	dir := filepath.Join(home, legacyProfiles)
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil
	}
	for _, entry := range entries {
		if !strings.HasSuffix(entry.Name(), ".json") || strings.HasPrefix(entry.Name(), ".") {
			continue
		}
		name := strings.TrimSuffix(entry.Name(), ".json")
		target := filepath.Join(home, name)
		if err := os.MkdirAll(target, 0o755); err != nil {
			return err
		}
		file := filepath.Join(target, constants.ProfileFile)
		if err := os.Rename(filepath.Join(dir, entry.Name()), file); err != nil {
			return err
		}
		for _, suffix := range []string{".key", ".key.pub"} {
			// A profile with no key is the ordinary case, so a rename that finds nothing is not
			// a failure.
			_ = os.Rename(filepath.Join(dir, name+suffix), filepath.Join(target, "profile"+suffix))
		}
		if err := repointKey(file, filepath.Join(dir, name+".key")); err != nil {
			return err
		}
	}
	return os.RemoveAll(dir)
}

// repointKey: the key moved with the profile, and the profile names it by absolute path.
func repointKey(file, was string) error {
	body, err := os.ReadFile(file)
	if err != nil {
		return nil
	}
	value, ok := jsjson.Parse(string(body))
	if !ok {
		return nil
	}
	held, isObject := value.(*jsjson.Object)
	if !isObject {
		return nil
	}
	if named, _ := held.Get("sshKeyPath"); named != any(was) {
		return nil
	}
	next := held.Clone()
	next.Set("sshKeyPath", strings.TrimSuffix(file, ".json")+".key")
	return os.WriteFile(file, []byte(jsjson.Indent(next, "  ")+"\n"), 0o600)
}

func writeProfile(home, name string) error {
	if err := os.MkdirAll(filepath.Join(home, name), 0o755); err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(home, name, constants.ProfileFile), []byte("{}\n"), 0o600)
}

// land moves the staged project into the profile that owns it. A name already taken there is
// only ever a migration that stopped halfway, and neither folder is worth losing to the other.
func land(from, profileDir string) (string, error) {
	if err := os.MkdirAll(profileDir, 0o755); err != nil {
		return "", err
	}
	name := filepath.Base(from)
	target := filepath.Join(profileDir, name)
	for n := 2; exists(target); n++ {
		target = filepath.Join(profileDir, name+"-"+strconv.Itoa(n))
	}
	if err := os.Rename(from, target); err != nil {
		return "", err
	}
	return target, nil
}

// rewrite is everything this machine filed under a project path, moved onto the path it now has.
func rewrite(held config.Config, paths map[string]string, names []string) config.Config {
	at := func(project string) string {
		if moved, found := paths[project]; found {
			return moved
		}
		return project
	}

	next := held.Clone()
	if held.ProjectPath != nil {
		moved := at(*held.ProjectPath)
		next.ProjectPath = &moved
	}
	next.Profile = profileFor(next.ProjectPath, held.Profile, names)
	next.Checkouts = rekey(held.Checkouts, at)
	next.Git = rekey(held.Git, at)
	next.Repo = rekey(held.Repo, at)
	next.RepoBranch = map[string]string{}
	for key, value := range held.RepoBranch {
		cut := strings.LastIndex(key, "#")
		if cut < 0 {
			next.RepoBranch[key] = value
			continue
		}
		next.RepoBranch[at(key[:cut])+key[cut:]] = value
	}
	return next
}

// profileFor: the open project sits inside the profile it commits as, so the path settles who
// that is. What the config said comes next, whether or not it names a profile that is there —
// and a home with neither falls back to whichever profile comes first.
func profileFor(projectPath, said *string, names []string) *string {
	if projectPath != nil {
		if held := pick(filepath.Base(filepath.Dir(*projectPath)), names); held != "" {
			return &held
		}
	}
	if said != nil {
		return said
	}
	if len(names) > 0 {
		return &names[0]
	}
	return nil
}

func rekey[T any](held map[string]T, at func(string) string) map[string]T {
	next := make(map[string]T, len(held))
	for key, value := range held {
		next[at(key)] = value
	}
	return next
}
